test/runtime/futex_test.c

```diff
@@ -0,0 +1,149 @@
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/syscall.h>
+#include <stdlib.h>
+#include <pthread.h>
+#include <runtime.h>
+
+#define EXIT_FAILURE 1
+#define EXIT_SUCCESS 0
+#define FUTEX_WAIT 0
+#define FUTEX_WAKE 1
+#define FUTEX_WAIT_BITSET 9
+#define FUTEX_INITIALIZER 0
+
+/* Test Global Variables */
+int wake_test_futex;
+int empty_futex;
+int wait_test_futex = FUTEX_INITIALIZER;
+int wait_bitset_test_futex = FUTEX_INITIALIZER;
+
+/* Helper Function Declarations */
+static void * futex_wake_test_thread(void *arg);
+
+/* FUTEX_WAKE test, creates num_to_wake threads which wait
+on uaddr and then wakes up all the threads */
+static boolean futex_wake_test(int* uaddr, int num_to_wake, int expected_result) {
+    pthread_t threads[num_to_wake];
+
+    for (int index = 0; index < num_to_wake; index++) {
+        if (pthread_create(&(threads[index]), NULL, futex_wake_test_thread, (void*)(uaddr))) {
+            printf("Unable to create thread\n");
+            return false;
+        }
+    }
+    
+    sleep(1); /* for main thread */
+    int ret = syscall(SYS_futex, uaddr, FUTEX_WAKE, num_to_wake, 0, NULL, 0);
+    if (ret == expected_result)
+        return true; /* success */
```

Comment:
Before returning, should `pthread_join()` all the threads, which would test that indeed all the threads have been woken up (or never went to sleep)

Reply:
Hello @francescolavra, does this mean to iterate over all the thread IDs and `pthread_join()` them? And then only return true if all the `pthread_join` calls are successful?

Reply:
Hello @francescolavra, to clarify, after iterating over all the threads and calling `pthread_join` on them, I should only return true if all the `pthread_join` calls are successful?

Reply:
Yes, only return true if (ret == expected_result) and all the pthread_join calls are successful